src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,365 @@
+            };
+
+            // Recursively sort descendant DocumentSymbolViewModels
+            foreach (var documentSymbolModel in sortedDocumentSymbolModels)
+            {
+                documentSymbolModel.Children = Sort(documentSymbolModel.Children, sortOption, oldSnapshot, currentSnapshot);
+            }
+
+            return sortedDocumentSymbolModels;
+        }
+
+        /// <summary>
+        /// Returns an immutable array of DocumentSymbolViewModels such that each node or one of its descendants matches the given pattern.
+        /// </summary>
+        public static ImmutableArray<DocumentSymbolViewModel> Search(ImmutableArray<DocumentSymbolViewModel> documentSymbolModels, string pattern)
+        {
+            var documentSymbols = ArrayBuilder<DocumentSymbolViewModel>.GetInstance();
+            var patternMatcher = PatternMatcher.CreatePatternMatcher(pattern, includeMatchedSpans: false, allowFuzzyMatching: true);
+
+            foreach (var documentSymbol in documentSymbolModels)
+            {
+                if (SearchNodeTree(documentSymbol, patternMatcher))
+                    documentSymbols.Add(documentSymbol);
+            }
+
+            return documentSymbols.ToImmutableAndFree();
+        }
+
+        /// <summary>
+        /// Returns true if the name of one of the tree nodes results in a pattern match.
+        /// </summary>
+        public static bool SearchNodeTree(DocumentSymbolViewModel tree, PatternMatcher patternMatcher)
+        {
+            if (patternMatcher.Matches(tree.Name))
+                return true;
+
+            foreach (var childItem in tree.Children)
+            {
+                if (SearchNodeTree(childItem, patternMatcher))
+                    return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Selects the Document Symbol node that is currently selected by the caret in the editor.
+        /// </summary>
+        /// <remarks>
+        /// The parameter lspSnapshot refers to the snapshot used when the LSP document symbol request was made
+        /// to obtain the symbol and currentSnapshot refers to the latest snapshot in the editor. These parameters
+        /// are required to obtain the latest DocumentSymbolViewModel range positions in the new snapshot.
+        /// </remarks>
+        public static void SelectDocumentNode(
+            ImmutableArray<DocumentSymbolViewModel> symbolTreeItemsSource,
+            ITextSnapshot currentSnapshot,
+            ITextSnapshot lspSnapshot,
+            int caretPosition)
+        {
+            LspSnapshot = lspSnapshot;
+            CurrentSnapshot = currentSnapshot;
+            CaretPosition = caretPosition;
+            IsSnapshotInitialized = true;
+
+            UnselectAll(symbolTreeItemsSource);
+            var selectedNode = GetNodeSelectedByCaret(symbolTreeItemsSource);
+            if (selectedNode is not null)
+                SelectNode(selectedNode);
```

Comment:
Can this just call SelectNode if we refactor it to take a list of elements (which could be eitehr the top list or the children)?